### PR TITLE
Add provider repositories on require command

### PR DESCRIPTION
**Pull request by aschempp**
Our Composer plugin dynamically registers repository URLs for packages in the artifact repository `/contao-manager/packages`. Currently, it only adds repositories for packages that are required (= are listed in the root `composer.json`). 

This pull request extends the capability to also register repository URLs in the `composer require` command. This fixes the problem that you cannot require a `contao-provider` and a package provided by that provider at the same time.

### CODE REVIEW

File: src/Composer/ArtifactsPlugin.php [hunk 1]
```diff
@@ -57,6 +74,28 @@ public function uninstall(Composer $composer, IOInterface $io): void
...
     }
 
+    public function preCommandRun(PreCommandRunEvent $event): void
+    {
+        if (!$this->composer || !$this->artifacts || 'require' !== $event->getCommand()) {
+            return;
+        }
+
+        $packages = $event->getInput()->getArgument('packages');
+        $requirements = (new \Composer\Package\Version\VersionParser())->parseNameVersionPairs($packages);
```
Comment (Toflar) [inline src/Composer/ArtifactsPlugin.php]: CS? FQCN?

File: tests/Composer/ArtifactsPluginTest.php [hunk 2]
```diff
@@ -480,6 +477,75 @@ public function testImplementsTheAPI2Methods(): void
...
     }
 
+    public function testRegistersPreCommandeEvent(): void
+    {
+        $this->assertSame([PluginEvents::PRE_COMMAND_RUN => ['preCommandRun', 1]], ArtifactsPlugin::getSubscribedEvents());
+    }
+
+    public function testRegistersContaoProvidersFromRequireCommand(): void
+    {
+        putenv('COMPOSER='.__DIR__.'/../Fixtures/Composer/artifact-data/composer.json');
```
Comment (Toflar) [inline tests/Composer/ArtifactsPluginTest.php:487]: Should be removed again after test run because of side effects.

**Review by Toflar — COMMENTED** [review]
LGTM, just minor nitpicks

File: src/Composer/ArtifactsPlugin.php (hunk 1 above)
Comment (aschempp) [reply src/Composer/ArtifactsPlugin.php]: There's already another `VersionParser` from semver required.

File: tests/Composer/ArtifactsPluginTest.php (hunk 2 above)
Comment (aschempp) [reply tests/Composer/ArtifactsPluginTest.php:487]: the `setUp` constant resets it. Not really related to this PR imho.

File: src/Composer/ArtifactsPlugin.php (hunk 1 above)
Comment (Toflar) [reply src/Composer/ArtifactsPlugin.php]: That one extends the one from Semver, I think you can use just the one from Composer?

Comment (aschempp) [reply src/Composer/ArtifactsPlugin.php]: good catch, didn't see that. Changed in 0fbf36c4a0992288daf588e258ace52328a987d5

**Review by Toflar — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @aschempp.

**Pull request merged**